### ISSUE TITLE
Support for handling popup windows

### ISSUE THREAD

**Issue by arsher** [MicrosoftEdge/WebViewFeedback#6]
I played around a little bit with the example and checked out the docs (https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2webview) but I couldn't find any way to handle popups. As far as I can tell it's currently not possible to open popups using window.open as it doesn't seem to do anything and returns null. So I'm guessing popups are currently disabled in this scenario (or I'm missing something :smile:).
It would be nice having some way to handle these as well, for example some kind of preview event to allow/disallow the opening of a popup and a way to initialize a new WebView instance for it. Do you have any plans for such a feature?

**Comment by david-risney** [comment]
Thanks for trying it out! Yes, a new window event is something we have in our plans as well as window.open working via script injection. To ensure our plans can satisfy your scenario, can you describe what you would use the event for? For instance, is it for popup windows related to auth scenarios?

**Comment by verelpode** [comment]
I agree, and I'd like to request that WebView2 support the same ["NewWindowRequested"](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webview.newwindowrequested) event as WebView 1.  Please note that some care needs to be taken in order to correctly handle temporary/session cookies, meaning cookies that have NO expiration date, or that have the "HTTP Only" option set to true.  When opening the URI in a new window in a new instance of WebView2, this only works correctly if the new instance of WebView2 has access to the same cookies.  For comparison, in the case of WebView 1, the scenario fails when you use the separate-process option and you try to open the new-window URI in a WebView that is running in a separate process.

Thus, when WebView2 supports the `NewWindowRequested` event, please ensure that it's possible and easy to give both instances of WebView2 (the original and the new window) access to the same set of temporary/session cookies.  This means that it is important to allow consumers of WebView2 to be able to:
* Easily select which [WebViewControlProcess](https://docs.microsoft.com/en-us/uwp/api/windows.web.ui.interop.webviewcontrolprocess) instance is used in each instance of WebView2.  (This affects which temporary/session cookies are available, and thus affects whether a new window/popup can be opened successfully or not.)
* Start up new instances of the web process.
* Explicitly instruct a web process instance to terminate, meaning the same as [WebViewControlProcess.Terminate](https://docs.microsoft.com/en-us/uwp/api/windows.web.ui.interop.webviewcontrolprocess.terminate).

**Comment by verelpode** [comment]
@david-risney 
> can you describe what you would use the event for? For instance, is it for popup windows related to auth scenarios?

In our case, the popup windows are not for entering a login/password (I presume that's what you mean by auth scenarios).  The popup windows are simply various different places in the websites where the website creators have decided to use popup windows for some reason.  Our group doesn't create the websites.  We're required to be compatible with the websites, and the websites happen to trigger popup webpages in various random places, therefore we must support it.

Furthermore, we're also required to support popups that show popups that show popups!  Although nested popups are a bit strange, nevertheless it works fine with the existing `NewWindowRequested` event in WebView 1.

If by "auth scenarios" you meant that the popup webpages are protected by cookies, then yes, we're definitely require to support popup webpages that must have access to the same cookies as the original webpage.  But no I haven't seen a login/password form in any of these popups, but I don't understand why that makes any difference, so maybe I misunderstood what you meant by "auth scenarios".

**Comment by liminzhu** [comment]
So a couple of clarification here. Pop-ups with window.open works with WebView2. 

![Capture](https://user-images.githubusercontent.com/1158954/63308731-de37bd80-c2a8-11e9-9655-d21f22f55d9d.PNG)
(this is from a WebView navigating to https://www.w3schools.com/jsref/met_win_open.asp and clicking the "Try it Yourself" button). 

The about-to-be-released `newWindowRequested` event allows the host to intercept and supply a new WebView window for the web content instead of letting a regular browser pop-up (like the one in the screenshot) to show up. As the host is in charge of supplying that WebView window, as long as that window is created from the same [user data folder](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/webview2.idl#createwebview2environmentwithdetails) as the WebView originated the popup, they will have access to the same cookies and what not with general the same origin rules applied.

**Comment by liminzhu** [comment]
Not super familiar with `*.HttpClient`, but they are definitely not implemented using the Chromium stack, so it'd be the equivalent of sharing cookies between different browsers which is not supported today and we'll have to think through it. 

> If an app cancels the new window and downloads the URI by itself (using Windows.Web.Http.HttpClient or System.Net.Http.HttpClient)

Can you elaborate more about why your app is doing this? Want to make sure I understand the scenario completely.

**Comment by verelpode** [comment]
@liminzhu
> it'd be the equivalent of sharing cookies between different browsers which is not supported today and we'll have to think through it.

I don't expect that it must be a fully automated/synchronized sharing feature.  In order to solve the cookie difficulties that we experience, it would be sufficient to make a "GetCookies" method in WebView2 that retrieves all of the cookies (both temporary and persistent cookies) that are stored in the user data folder, completely independent of `*.HttpClient`.  I would then be able to bridge the gap between WebView2 and HttpClient by writing code that invokes `WebView2.GetCookies` followed by [Windows.Web.Http.HttpCookieManager.SetCookie](https://docs.microsoft.com/en-us/uwp/api/windows.web.http.httpcookiemanager.setcookie).  Thus WebView2 would allow me to retrieve the cookies, and then I make those same cookies available to `HttpClient` via `HttpCookieManager`.

In addition to "GetCookies", it would also be good to have a "SetCookies" method in WebView2 in order to support transferring of cookies in the reverse direction, meaning from HttpClient to WebView2.

> Can you elaborate more about why your app is doing this?

WebView2 cannot display all content types / file types.  Our app displays what WebView2 cannot display.  Our app uses `*.HttpClient` to download the content prior to displaying it.  The catch is that the website rejects the request if the session cookies are missing.  Thus the need to retrieve the cookies from WebView2 and insert them into HttpClient.  In the case of WebView 1, our app subscribes to the event [UnviewableContentIdentified](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webview.unviewablecontentidentified) and proceeds to use `HttpClient` to request/download the URI specified in the property [WebViewUnviewableContentIdentifiedEventArgs.Uri](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webviewunviewablecontentidentifiedeventargs.uri), but at this point it fails if the cookies are missing.

The second reason is to download large files that WebView2 cannot or should not download.  Firstly, WebView2 displays a webpage that contains a clickable link to a file.  The user clicks the link.  Via events, our app can respond by using `HttpClient` to download the linked file.  Again this requires retrieving the cookies from WebView2 and inserting them into `HttpClient`/`HttpCookieManager`.

See also [issue #4](https://github.com/MicrosoftEdge/WebViewFeedback/issues/4).

**Comment by liminzhu** [comment]
@verelpode thanks for the context! I think I'm getting a full picture of what you're trying to achieve now :) You are invoking HttpClient to download unviewable content and display it in your own viewer and cookies need to be there to persistent a logged in session otherwise you might not be able to download the resource. 

Yeah I think this comes back to the cookie management request in https://github.com/MicrosoftEdge/WebViewFeedback/issues/4 which is definitely on our backlog.

**Comment by verelpode** [comment]
@liminzhu
> You are invoking HttpClient to download unviewable content and display it in your own viewer and cookies need to be there to persistent a logged in session otherwise you might not be able to download the resource.

Exactly.  The unviewable content may come via a `NewWindowRequested` or other paths such as simple normal links.  This is already causing a _major_ problem for us because we're unable to deploy the new software version outside of the lab because essential websites reject the downloads because the session cookies are missing because the current version of WebView-1-for-WPF (and also WebView2) can only run in a separate process thereby making the session cookies unavailable to our app/process that is trying to use HttpClient to download the unviewable content and display it in our own viewers.  The required cookies are _doubly_ unavailable when using WebView2 because not only is it a separate process but it also doesn't use WinINet.dll anymore.

The download + display of unviewable content is trigged via multiple paths including:
1. The `NewWindowRequested` event, when [WebViewNewWindowRequestedEventArgs.Uri](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webviewnewwindowrequestedeventargs.uri) is a URI that responds with an unviewable content-type.
2. When [IWebView2NavigationStartingEventArgs.Uri](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2navigationstartingeventargs#get_uri) is a URI that responds with an unviewable content-type.
3. When the HTML `<embed src="http://...">` tag specifies a URI that responds with an unviewable content-type.
4. When the HTML `<iframe src="http://...">` tag is used and the inline frame is navigated to a URI that responds with an unviewable content-type.  Event [FrameNavigationStarting](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2webview#add_framenavigationstarting).
5. Possibly also via some JavaScript.

If you make a `UnviewableContentIdentified` event in WebView2, it would be helpful if the event args can specify the origin of the unviewable content, meaning:  Via `NewWindowRequested`, or a normal hyperlink/anchor, or `<embed>`, or an inline frame, or JavaScript.

**Comment by liminzhu** [comment]
As I'm thinking through this, how to do plug in and position your own viewer for embedded resource (`<embed ...>`/`<iframe ...>`)? If you're navigating to a uri that ends up being unviewable I think you probably just size and place your own viewer on top of the WebView?

**Comment by verelpode** [comment]
I posted my reply over in #38.  Currently with WebView 1, we show a viewer in response to `UnviewableContentIdentified` and this works well with normal clickable links, and it also works in conjunction with `NewWindowRequested`, but it doesn't work properly with `<embed ...>` or `<iframe ...>`.

**Comment by verelpode** [comment]
Thanks for creating the [NewWindowRequested](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2webview3#add_newwindowrequested) event; that's great news!  I especially like the new [IsUserInitiated](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2newwindowrequestedeventargs#get_isuserinitiated) property.  A few questions:

### Referrer
[IWebView2NewWindowRequestedEventArgs](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2newwindowrequestedeventargs#iwebview2newwindowrequestedeventargs) does not contain an equivalent of [Windows.UI.Xaml.Controls.WebViewNewWindowRequestedEventArgs.Referrer](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webviewnewwindowrequestedeventargs.referrer).  Is there a reason for removing the `Referrer` property?  I suggest creating the same `Referrer` property in WebView2, if it's not excessively difficult to create.  On the other hand, admittedly so far I've only ever used that `Referrer` property for diagnostics/troubleshooting purposes.  I suppose it's a case of useful info but not strictly essential.

### URI Query and Fragment
I suggest updating the documentation to explain whether the [IWebView2NewWindowRequestedEventArgs.Uri](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2newwindowrequestedeventargs#get_uri) property includes or excludes the URI's Query and Fragment parts.
* Query meaning when the URI contains "?" meaning [System.Uri.Query](https://docs.microsoft.com/en-us/dotnet/api/system.uri.query).
* Fragment meaning when the URI contains "#" meaning [System.Uri.Fragment](https://docs.microsoft.com/en-us/dotnet/api/system.uri.fragment).

I remember that somewhere someone reported a problem where the Query or Fragment part of the URI was lost, but I can't remember whether it was the Query or Fragment, and I can't remember whether it was WebView2 or EdgeHTML WebView.  I suggest testing and documentation on these points.

### Same environment/folder/process?
When setting the property [IWebView2NewWindowRequestedEventArgs.NewWindow](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2newwindowrequestedeventargs#put_newwindow) to a new WebView2 instance, will an error code be reported somewhere if the WebView2 instance uses a different environment or user-data-folder or process?  I suggest adding a sentence to explain this point in the documentation.

If it accepts a different environment without erroring, my understanding is that in practice the same environment or user-data-folder is still necessary in order to give the new "window" access to the same session cookies as the original "window".

### Reasoning for the NewWindow property
In comparison, EdgeHTML [Windows.UI.Xaml.Controls.WebViewNewWindowRequestedEventArgs](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.webviewnewwindowrequestedeventargs) does not contain an equivalent of the property [IWebView2NewWindowRequestedEventArgs.NewWindow](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2newwindowrequestedeventargs#put_newwindow).   Why does WebView2 have the `NewWindow` property when EdgeHTML-WebView didn't?  Can you explain this in the documentation?  Actually I think the `NewWindow` property is a good idea but today I've forgotten the reason for it. Maybe I'll remember tomorrow.  Can you explain it?

In other words:  Why can't we create a new instance of WebView2 using the same environment and user-data-folder and then invoke `IWebView2WebView.Navigate` to navigate the new WebView2 to the URI specified in `IWebView2NewWindowRequestedEventArgs.Uri`?  How is that different to setting the `IWebView2NewWindowRequestedEventArgs.NewWindow` property?

**Comment by david-risney** [comment]
* **Referer**: Yes good point. It wasn't simple to expose and we didn't have a customer that required it. We can look at adding it in the future if the need comes up.

 * **URI**: There was a bug with EdgeHtml WebView where the fragment was missing from the URI for the new window event. Generally I would want to assume that the URI is complete and document when it isn't. 

 * **NewWindow**: Good questions. I'll open a separate issue to improve the NewWindow documentation to answer the questions:
   * Is it supported to provide a webview from a different environment?
   * If not, what happens when one is provided?
   * Include what happens for all combinations of NewWindow is set or not and Handled is set or not. What happens if the NewWindow property isn't set and handled isn't set? (A browser-like new window is created for you)
   * Describe why the NewWindow property is useful. (A page can programmatically open a new window, retain a reference to that new window in script, and interact with that new window script reference eg `let w = window.open("https://example.com/"); w.postMessage(...)`. Marking the event handled, not providing a webview, and then separately creating your own webview means the new webview won't be attached to that window.open return value)

**Comment by sandrohanea** [comment]
Hi all!
Should this event(NewWindowRequested) work for Ctrl + Click as well?
I tested it now and it seems it's not firing for this case. Is it possible to block the user to not navigate away using Ctrl+Click?

**Comment by champnic** [comment]
Yes this should work for Ctrl+Click as well. This is a known bug and we are currently fixing it. Please see #371 for tracking.

**Comment by champnic** [comment]
I'm going to Close this issue for now, as I believe the NewWindowRequested event covers it. If there's something missing please reopen or open a new issue. Thanks!

**Comment by Symbai** [comment]
Where do I find this "NewWindowRequested" event?! In VS WPF this event does not exist:

![image](https://user-images.githubusercontent.com/14368203/92763205-96904200-f393-11ea-867a-780f94c812c6.png)

**Comment by champnic** [comment]
It's an event on CoreWebView2, not the WebView2 control itself. See https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/dotnet/0-9-538/microsoft-web-webview2-core-corewebview2#newwindowrequested. Your code should handle the CoreWebView2Ready event, and then do webView.CoreWebView2.NewWindowRequested += ...

**Comment by dec** [comment]
Hello to all,

Apparently the "NewWindowRequested" event is not fired when we use the middle mouse button in a link... at least this is the behaviour that I am experimenting here...

**Comment by seikosantana** [comment]
I need to prevent the app creating any window, including opening downloaded files. Can we override the file opening and open it on default opener instead of letting it creating a window? NewWindowRequested isn't fired on that too.

**Comment by joinshoaib** [comment]
> Thanks for trying it out! Yes, a new window event is something we have in our plans as well as window.open working via script injection. To ensure our plans can satisfy your scenario, can you describe what you would use the event for? For instance, is it for popup windows related to auth scenarios?

i used for auth scenario.. how i can implement??
